src/solidity/contracts.js

```diff
@@ -0,0 +1,167 @@
+    for (var k in stateVars) {
+      var variable = stateVars[k]
+      if (variable.name === 'VariableDeclaration') {
+        var type = varUtil.getType(variable, stateVars)
+        var loc = varUtil.walkStorage(type, location)
+        ret.push({
+          name: variable.attributes.name,
+          type: type,
+          location: loc.currentLocation
+        })
+        location = loc.nextLocation
+      }
+    }
+    return ret
+  },
+
+  /**
+   * return state var and type definition of the given contract
+   *
+   * @param {Object} astList  - AST nodes of all the sources
+   * @param {String} contractName  - contract for which state var should be resolved
+   * @return {Object} - return the AST node of found state var
+   */
+  getStateDefinition: function (astList, contractName) {
+    var id = this.getContractId(astList, contractName)
+    if (id) {
+      var stateVar = []
+      var ctrs = this.getContractsDefinition(astList)
+      var baseContracts = this.getBaseContracts(astList, id, ctrs)
+      for (var k in baseContracts) {
+        var ctr = baseContracts[k]
+        for (var i in ctr.children) {
+          stateVar.push(ctr.children[i])
+        }
+      }
+      return stateVar
+    }
+    return null
+  },
+
+  /**
+   * return contract id given his name
+   *
+   * @param {Object} astList  - AST nodes of all the sources
+   * @param {String} contractName  - contract for which based contracts should be resolved
+   * @return {Object} - return the AST node of found contract
+   */
+  getContractId: function (astList, contractName) {
+    var id
+    this.walkAstList(astList, { 'ContractDefinition': function (node) {
+      if (node.attributes && node.attributes.name === contractName) {
+        id = node.id
+        return false
+      } else {
+        return true
+      }
+    }})
+    return id
+  },
+
+  /**
+   * return all the based contracts defined given @arg contract name
+   *
+   * @param {Object} astList  - AST nodes of all the sources
+   * @param {Int} id  - contract id for which based contracts should be resolved
+   * @param {Map} ctrs  - all contracts defined in the current context
+   * @return {Object} - return the AST node of found contract
+   */
+  getBaseContracts: function (astList, id, ctrs) {
+    var contract = ctrs[id]
+    var baseContracts = []
+    for (var k in contract.attributes.linearizedBaseContracts) {
+      var subId = contract.attributes.linearizedBaseContracts[k]
+      baseContracts.unshift(ctrs[subId])
+    }
+    return baseContracts
+  },
+
+  /**
+   * return all contract definitions of the given @astList
+   *
+   * @param {Object} astList  - AST nodes of all the sources
+   * @return {Object} - return the AST node of found contract
+   */
+  getContractsDefinition: function (astList) {
+    var callback = {}
+    var ctrs = {}
+    callback.ContractDefinition = function (node) {
+      ctrs[node.id] = node
+      return false
+    }
+    this.walkAstList(astList, callback)
```

Comment:
Does it make sense to extract all the information we need with a single walk of the AST?